### PR TITLE
chore: publish test reports to github pages [WD-7907]

### PR DESCRIPTION
**Pull request by mas-who**
## Done

- Created an orphan branch gh-pages from which github pages for this repository will deploy
- Added a new job in pr.yaml to use the artifact produced by the e2e test job and push it to the gh-pages branch. This takes into consideration that multiple pr actions may be running at the same time. Lastly, the job will run if the e2e tests failed or succeeded and output a url for the test reports.
- Added a new action triggered on PR merge so that test reprorts gets cleaned up on the gh-pages branch.

### CODE REVIEW

File: .github/workflows/delete_test_reports.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+              if: ${{ steps.has-reports.outputs.BRANCH_REPORTS_EXIST == 'true' }}
+              timeout-minutes: 3
+              run: |
+                cd $BRANCH_REPORTS_DIR/..
+        
+                rm -rf ${{ github.event.ref }}
+                git add .
+                git commit -m "workflow: remove all reports for branch ${{ github.event.ref }}"
+        
+                while true; do
```
Comment (edlerd) [inline .github/workflows/delete_test_reports.yml:39]: Why the while loop?

File: .github/workflows/delete_test_reports.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                else
+                    echo "BRANCH_REPORTS_EXIST=true" >> $GITHUB_OUTPUT
+                fi
+            - name: Delete reports from repo for branch
+              if: ${{ steps.has-reports.outputs.BRANCH_REPORTS_EXIST == 'true' }}
+              timeout-minutes: 3
+              run: |
+                cd $BRANCH_REPORTS_DIR/..
+        
+                rm -rf ${{ github.event.ref }}
```
Comment (edlerd) [inline .github/workflows/delete_test_reports.yml:35]:
```suggestion
                rm -rf reports/${{ github.event.ref }}
```
Simpler?

File: .github/useful_act_commands.md
Comment (edlerd) [inline .github/useful_act_commands.md:1]: Why is this file needed?

File: .github/workflows/pr.yaml [hunk 3]
```diff
@@ -155,3 +157,53 @@ jobs:
...
+          name: playwright-report
+          path: ${{ env.HTML_REPORT_URL_PATH }}
+      - name: Push HTML Report
+        timeout-minutes: 3
+        # In case if another action job pushed to gh-pages while we are rebasing for the current job
+        run: |
+          git add .
+          git commit -m "workflow: add HTML report for branch ${{ github.head_ref || github.ref_name }} (run_id: ${{ github.run_id }}, attempt:  ${{ github.run_attempt }})"
+
+          while true; do
```
Comment (edlerd) [inline .github/workflows/pr.yaml:193]: Why the loop?

File: .github/workflows/pr.yaml [hunk 4]
```diff
@@ -155,3 +157,53 @@ jobs:
...
+
+  publish_report:
+    name: publish-e2e-report
+    # if job is cancelled we should skip this
+    if: success() || needs.e2e.outputs.job_status == 'failure'
+    needs: [e2e]
+    runs-on: ubuntu-latest
+    continue-on-error: true
+    env:
+      HTML_REPORT_URL_PATH: reports/${{ github.head_ref || github.ref_name }}/${{ github.run_id }}/${{ github.run_attempt }}
```
Comment (edlerd) [inline .github/workflows/pr.yaml:169]: So we keep multiple runs on the same run available, nice :+1:

File: .github/workflows/delete_test_reports.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                        echo "Failed to rebase. Please review manually."
+                        exit 1
+                    fi
+        
+                    git push
+                    if [ $? -eq 0 ]; then
+                        echo "Successfully pushed HTML reports to repo."
+                        exit 0
+                    fi
+                done
```
Comment (edlerd) [inline .github/workflows/delete_test_reports.yml:51]: Let's add a newline 
```suggestion
                done

```

**Review by edlerd — COMMENTED** [review]
This is quite nice. Some questions and ideas below. Happy to discuss them over a call.

File: .github/workflows/pr.yaml (hunk 4 above)
Comment (edlerd) [reply .github/workflows/pr.yaml:169]: Maybe use the PR number instead of the branch name? Then we can trigger the deletion on PR merge instead of branch removal. Just an idea, maybe simpler to keep as it is.

File: .github/useful_act_commands.md
Comment (mas-who) [reply .github/useful_act_commands.md:1]: I wanted to keep a record of nektos commands for future reference, I will move this out of the repo.

File: .github/workflows/delete_test_reports.yml (hunk 1 above)
Comment (mas-who) [reply .github/workflows/delete_test_reports.yml:39]: This is to cater for the scenario during a rebase, another pr action job pushed to the gh-pages branch. This means the push on the current job run will fail. If that's the case and the exit code is non-zero, then we will rebase again and push. Does that make sense or do you see some flaws in the logic?

File: .github/workflows/pr.yaml (hunk 4 above)
Comment (mas-who) [reply .github/workflows/pr.yaml:169]: I don't see too much of a functional difference between tracking by PR number or branch name since this job is triggered with a pull request so it has to be an PR for a specific branch. The report delete action first checks if a folder with the branch name exists before deleting anything, so I don't think there is a risk there. Changing the implementation shouldn't be too bad but I do need to go through another round of testing which could take some time. If you think there is a definite benefit in doing this change then I can make it happen?

File: .github/workflows/pr.yaml (hunk 3 above)
Comment (mas-who) [reply .github/workflows/pr.yaml:193]: same comment as before

File: .github/workflows/pr.yaml (hunk 4 above)
Comment (edlerd) [reply .github/workflows/pr.yaml:169]: I just realized: The source branch can be in a fork, and then the current workflow would miss the "delete branch" event. So I think we have to track PRs and clean up when they are closed.

File: .github/workflows/delete_test_reports.yml (hunk 1 above)
Comment (edlerd) [reply .github/workflows/delete_test_reports.yml:39]: Makes sense, maybe add the comment above the loop?

File: .github/workflows/pr.yaml (hunk 3 above)
Comment (edlerd) [reply .github/workflows/pr.yaml:193]: Makes sense, maybe also add the comment in the line above the loop?

Comment (mas-who) [reply .github/workflows/pr.yaml:193]: sure done

File: .github/workflows/delete_test_reports.yml (hunk 1 above)
Comment (mas-who) [reply .github/workflows/delete_test_reports.yml:39]: I don't think you can comment in a run command, maybe I just echo an explanation?

File: .github/workflows/pr.yaml (hunk 4 above)
Comment (edlerd) [reply .github/workflows/pr.yaml:169]: On the cleanup job, you can use 
```
on:
  pull_request:
    types: [ closed ]
```

Comment (mas-who) [reply .github/workflows/pr.yaml]: the actions are now tracking by PR number.

Screenshot from 2023-12-13 17-09-53

File: .github/workflows/delete_test_reports.yml (hunk 1 above)
Comment (edlerd) [reply .github/workflows/delete_test_reports.yml:39]: It should work if you start the line with a `#`, as that is a comment in a shell as well.

**Review by edlerd — APPROVED** [review]
LGTM, just one nit about the commenting of the while loops. Feel free to merge as is though.

**Comment by mas-who** [comment]
> LGTM, just one nit about the commenting of the while loops. Feel free to merge as is though.

Aaah right, it's a shell script actually so it should be fine. I've moved the comment to within the run command on both actions now.
